src/library/scala/collection/mutable/RedBlackTree.scala

```diff
@@ -0,0 +1,502 @@
+          }
+          z.parent.red = false
+          z.parent.parent.red = true
+          rotateLeft(tree, z.parent.parent)
+        }
+      }
+    }
+    tree.root.red = false
+  }
+
+  // ---- deletion ----
+
+  def delete[A, B](tree: Tree[A, B], key: A)(implicit ord: Ordering[A]): Unit = {
+    val z = getNode(tree.root, key)
+    if (z ne null) {
+      var y = z
+      var yIsRed = y.red
+      var x: Node[A, B] = null
+      var xParent: Node[A, B] = null
+
+      if (z.left eq null) {
+        x = z.right
+        transplant(tree, z, z.right)
+        xParent = z.parent
+      }
+      else if (z.right eq null) {
+        x = z.left
+        transplant(tree, z, z.left)
+        xParent = z.parent
+      }
+      else {
+        y = minNodeNonNull(z.right)
+        yIsRed = y.red
+        x = y.right
+
+        if (y.parent eq z) xParent = y
+        else {
+          xParent = y.parent
+          transplant(tree, y, y.right)
+          y.right = z.right
+          y.right.parent = y
+        }
+        transplant(tree, z, y)
+        y.left = z.left
+        y.left.parent = y
+        y.red = z.red
+      }
+
+      if (!yIsRed) fixAfterDelete(tree, x, xParent)
+      tree.size -= 1
+    }
+  }
+
+  def fixAfterDelete[A, B](tree: Tree[A, B], node: Node[A, B], parent: Node[A, B]): Unit = {
+    var x = node
+    var xParent = parent
+    while ((x ne tree.root) && isBlack(x)) {
+      if (x eq xParent.left) {
+        var w = xParent.right
+        // assert(w ne null)
+
+        if (w.red) {
+          w.red = false
+          xParent.red = true
+          rotateLeft(tree, xParent)
+          w = xParent.right
+        }
+        if (isBlack(w.left) && isBlack(w.right)) {
+          w.red = true
+          x = xParent
+        } else {
+          if (isBlack(w.right)) {
+            w.left.red = false
+            w.red = true
+            rotateRight(tree, w)
+            w = xParent.right
+          }
+          w.red = xParent.red
+          xParent.red = false
+          w.right.red = false
+          rotateLeft(tree, xParent)
+          x = tree.root
+        }
+      } else { // symmetric cases
+        var w = xParent.left
+        // assert(w ne null)
+
+        if (w.red) {
+          w.red = false
+          xParent.red = true
+          rotateRight(tree, xParent)
+          w = xParent.left
+        }
+        if (isBlack(w.right) && isBlack(w.left)) {
+          w.red = true
+          x = xParent
+        } else {
+          if (isBlack(w.left)) {
+            w.right.red = false
+            w.red = true
+            rotateLeft(tree, w)
+            w = xParent.left
+          }
+          w.red = xParent.red
+          xParent.red = false
+          w.left.red = false
+          rotateRight(tree, xParent)
+          x = tree.root
+        }
+      }
+      xParent = x.parent
+    }
+    if (x ne null) x.red = false
+  }
+
+  // ---- helpers ----
+
+  private[this] def successor[A, B](node: Node[A, B]): Node[A, B] = {
```

Comment:
I'll do that too. I think it suffices to say that it is the node that succeeds the argument `node` in an in-order traversal, do you think it also needs examples?